pandas/core/indexes/period.py

```diff
@@ -165,90 +198,212 @@ def __new__(cls, data=None, ordinal=None, freq=None, start=None, end=None,
-        result = super(PeriodIndex, cls)._simple_new(values, freq)
+            # PeriodIndex allow PeriodIndex(period_index, freq=different)
+            # Let's not encourage that kind of behavior in PeriodArray.
+
+            if freq and isinstance(data, cls) and data.freq != freq:
+                # TODO: We can do some of these with no-copy / coercion?
+                # e.g. D -> 2D seems to be OK
+                data = data.asfreq(freq)
+
+            data = period_array(data=data, ordinal=ordinal, freq=freq,
+                                copy=copy)
+
+        if copy:
+            data = data.copy()
+
+        return cls._simple_new(data, name=name)
 
+    # ------------------------------------------------------------------------
+    # Data
+    @property
+    def _ndarray_values(self):
+        return self.values.values
+
+    @property
+    def values(self):
+        return self._data
+
+    @property
+    def freq(self):
+        # TODO(DatetimeArray): remove
+        # Can't simply use delegate_names since our base class is defining
+        # freq
+        return self._data.freq
+
+    @freq.setter
+    def freq(self, value):
+        value = Period._maybe_convert_freq(value)
+        msg = ('Setting {cls}.freq has been deprecated and will be '
+               'removed in a future version; use {cls}.asfreq instead. '
+               'The {cls}.freq setter is not guaranteed to work.')
+        warnings.warn(msg.format(cls=type(self).__name__),
+                      FutureWarning, stacklevel=2)
+        # PeriodArray._freq isn't actually mutable. We set the private _freq
+        # here, but people shouldn't be doing this anyway.
+        self._data._freq = value
+
+    # ------------------------------------------------------------------------
+    # Index Constructors
+
+    @classmethod
+    def _simple_new(cls, values, name=None, freq=None, **kwargs):
+        # type: (PeriodArray, Any, Any) -> PeriodIndex
+        """
+        Values can be any type that can be coerced to Periods.
+        Ordinals in an ndarray are fastpath-ed to `_from_ordinals`
+        """
+        if isinstance(values, cls):
+            # TODO: don't do this
+            values = values.values
+        elif (isinstance(values, (ABCIndexClass, np.ndarray)) and
+                is_integer_dtype(values)):
+            # TODO: don't do this.
+            values = PeriodArray._simple_new(values, freq)
+
+        if not isinstance(values, PeriodArray):
+            raise TypeError("PeriodIndex._simple_new only accepts PeriodArray")
+        result = object.__new__(cls)
+        result._data = values
         result.name = name
         result._reset_identity()
         return result
 
-    def _shallow_copy_with_infer(self, values, **kwargs):
+    @classmethod
+    def _from_ordinals(cls, values, name=None, freq=None, **kwargs):
+        """
+        Values should be int ordinals
+        `__new__` & `_simple_new` cooerce to ordinals and call this method
```

Comment:
I think this method can be removed. It is only used in `_new_PeriodIndex`, where it can be replaced with combination of PeriodArray and simple_new I think). 
Also, docstring is not up to date.

Reply:
yes i agree

Reply:
+1